### ISSUE TITLE
Autocomplete doesn't recognize complex type's fields

### ISSUE THREAD

**Issue by enbyted** [Microsoft/vscode#28804]
- VSCode Version: 1.13.1
- OS Version: Windows 8.1 64-bit

I have the following type definitions:
```ts
interface DataType {
    key: string,
}
type PlatformType = 'hiperreg' | 'uC';
type PlatformDTO = PlatformHiperreg | PlatformMicroController;

interface PlatformBase extends DataType {
    type: PlatformType,
    name: string,
    supportsVersions: boolean,
}

interface PlatformHiperreg extends PlatformBase {
    type: 'hiperreg',
}

interface PlatformMicroController extends PlatformBase {
    type: 'uC',
    brand: 'avr' | 'pic',
    model: string,
}
```

and then I use it as follows:
```ts
const platforms: PlatformDTO[] = [
    {
        type: 'hiperreg',
        key: 'hiperreg',
        name: 'HiperReg',
        supportsVersions: true,
    },
    {
        type: 'uC',
        key: 'avr-atmega8',
        name: 'ATmega8',
        brand: 'avr',
        model: 'ATmega8',
        supportsVersions: false,
    },
]
```

What happens is then when I'm entering entry with `type: 'uC'` the autocomplete doesn't suggest fields `brand` and `model` which exists only on type `PlatformMicroController` which is distinguished by field `type`.

**Comment by mjbvz** [comment]
I believe this was fixed by TypeScript 2.4 which we just picked up in vscode insiders

<img width="839" alt="screen shot 2017-06-16 at 2 19 45 pm" src="https://user-images.githubusercontent.com/12821956/27245174-e35cfd2e-529e-11e7-8a83-1fdda0ade0a1.png">

Please give the latest insiders builds a try and let me know if you still see this issue: https://code.visualstudio.com/insiders

**Comment by enbyted** [comment]
@mjbvz Thanks, it's indeed working in the Insiders build, however since posting this issue I've changed the schema a bit and ~encountered another, but similar problem~ learned that for TypeScript to recognize types correctly you have to _expand_ and not _narrow_ them in derived types.
